src/output/plugins/visualization/Protocol.cxx

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The Music Player Daemon Project
+
+#include "Protocol.hxx"
+#include <netinet/in.h>
+
+#include "Log.hxx"
+#include "util/Domain.hxx"
+
+Visualization::ParseResult
+Visualization::ParseClihlo(void *data, size_t length, ClientHello &clihlo) noexcept {
+	// CLIHLO payload is 6 bytes, header & footer are five more.
+	if (length < sizeof(ClientHello) + 5) {
+		return ParseResult::NEED_MORE_DATA;
+	}
+
+	uint8_t *buf = (uint8_t *)data;
+
+	uint16_t msg_type = ntohs(*(uint16_t *)buf);
+	if (msg_type != 0) {
+		return ParseResult::ERROR;
+	}
+
+	buf += 2;
+	uint16_t payload_len = ntohs(*(uint16_t *)buf);
+	if (payload_len != 6) {
+		return ParseResult::ERROR;
+	}
+
+	buf += 2;
+	clihlo.major_version = *buf++;
+	clihlo.minor_version = *buf++;
+
+	clihlo.requested_fps = ntohs(*(uint16_t *)(buf));
+	buf += 2;
+	clihlo.tau = ntohs(*(int16_t *)(buf));
+	buf += 2;
+
+	if (*buf != 0) {
+		return ParseResult::ERROR;
+	}
+
+	return ParseResult::OK;
```

Comment:
All of this fragile could be eliminated if `struct ClientHello` were a big-endian struct, e.g. by using `PackedBE16` instead of `uint16_t`.

Reply:
Sorry-- could you say more? If I replaced the `uint16_t` members of `struct ClientHello` with `PackedBE16` equivalents, that would help how, exactly. I mean, I see that the calls to `ntohs` would go away, to be replaced by invocations of their `uint16_t()` methods, I suppose?

Reply:
What methods are invoked?